ngraph/frontend/onnx_import/src/op/org.openvinotoolkit/group_norm.cpp

```diff
@@ -0,0 +1,148 @@
+                    std::shared_ptr<ngraph::Node>
+                        create_group_norm_shape(const Output<ngraph::Node>& data, size_t num_groups)
+                    {
+                        const auto& pshape = data.get_partial_shape();
+                        NGRAPH_CHECK(pshape.rank().is_static());
+                        size_t rank_size = pshape.rank().get_length();
+                        NGRAPH_CHECK(rank_size >= 3, "3-D and above tensors supported only");
+
+                        if (pshape.is_static())
+                        {
+                            const auto& shape = pshape.to_shape();
+                            std::vector<size_t> new_shape{
+                                shape[0], num_groups, shape[1] / num_groups};
+                            for (size_t i = 2; i < rank_size; i++)
+                            {
+                                new_shape.push_back(shape[i]);
+                            }
+                            return default_opset::Constant::create(
+                                element::i64, Shape{new_shape.size()}, new_shape);
+                        }
+
+                        auto shape = std::make_shared<default_opset::ShapeOf>(data);
+                        auto splits = builder::opset1::split(shape, rank_size);
+                        auto num_groups_const =
+                            default_opset::Constant::create(element::i64, Shape{1}, {num_groups});
+                        NodeVector new_shape{
+                            splits[0].get_node_shared_ptr(),
+                            num_groups_const,
+                            std::make_shared<default_opset::Divide>(splits[1], num_groups_const)};
+                        for (size_t i = 2; i < rank_size; i++)
+                        {
+                            new_shape.push_back(splits[i].get_node_shared_ptr());
+                        }
+                        return std::make_shared<default_opset::Concat>(new_shape, 0);
+                    }
+                }
+            } // detail
+
+            namespace set_1
+            {
+                OutputVector group_norm(const Node& node)
+                {
+                    auto inputs = node.get_ng_inputs();
+                    NGRAPH_CHECK(inputs.size() == 3,
+                                 "Invalid number of inputs. Expected 3, actual " +
+                                     std::to_string(inputs.size()));
+
+                    auto data = inputs[0];
+                    auto scale = inputs[1];
+                    auto bias = inputs[2];
+
+                    size_t num_groups =
+                        static_cast<size_t>(node.get_attribute_value<int64_t>("num_groups"));
+                    float eps = node.get_attribute_value<float>("eps", 1e-5);
+
+                    auto data_pshape = data.get_partial_shape();
+                    std::shared_ptr<ngraph::Node> data_shape_node;
+                    if (data_pshape.is_static())
```

Comment:
I don't think it is correct. Even if shape is static we may want to reshape a model later before loading an nGraph function to a plugin. We have to keep shape sub-expressions as many as possible to maintain this capability.

Reply:
Take a look at group norm paper https://arxiv.org/pdf/1803.08494.pdf. By its definition, you have to apply reshape to the input.

Reply:
That's also what MO does.

Reply:
I'm not arguing that such reshape is not needed. I think that we always should follow the approach where ShapeOf is used and shouldn't use shape from partial_shape of a node even if it is statically known during the conversion. MO uses it at https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/middle/GroupNorm.py#L60. This code looks to be used unconditionally.